# Introducing to problem solving with pandas

### Dataset: "2_taxi_nyc.csv"

### Tasks
* Выяснить, из какого района было совершено наибольшее количество поездок за весь период
* Сравните среднее число поездок, и выберите районы, из которых по праздникам в среднем поступает больше заказов, чем в обычные дни
* Для каждого района посчитайте число поездок по месяцам. Отсортируйте полученные значения по убыванию и сохраните результирующий датафрейм в *pickups_by_mon_bor*
* Напишите функцию *temp_to_celcius*, которая получает на вход колонку с температурой в °F, и возвращает значения, переведенные в градусы Цельсия


### Found methods while solving tasks
**.isnull()**

**.idxmin()** 

**.idxmax()**

**.min()** 

**.max()**

In [1]:
import pandas as pd

In [2]:
taxi = pd.read_csv('2_taxi_nyc.csv')

In [3]:
taxi

pickup_dt pickup_month        borough  pickups hday  spd  \
0      2015-01-01 01:00:00          Jan          Bronx      152    Y  5.0   
1      2015-01-01 01:00:00          Jan       Brooklyn     1519    Y  5.0   
2      2015-01-01 01:00:00          Jan            EWR        0    Y  5.0   
3      2015-01-01 01:00:00          Jan      Manhattan     5258    Y  5.0   
4      2015-01-01 01:00:00          Jan         Queens      405    Y  5.0   
...                    ...          ...            ...      ...  ...  ...   
29096  2015-06-30 23:00:00          Jun            EWR        0    N  7.0   
29097  2015-06-30 23:00:00          Jun      Manhattan     3828    N  7.0   
29098  2015-06-30 23:00:00          Jun         Queens      580    N  7.0   
29099  2015-06-30 23:00:00          Jun  Staten Island        0    N  7.0   
29100  2015-06-30 23:00:00          Jun            NaN        3    N  7.0   

        vsb  temp  dewp     slp  pcp 01  pcp 06  pcp 24   sd  
0      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
1      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
2      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
3      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
4      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
...     ...   ...   ...     ...     ...     ...     ...  ...  
29096  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  
29097  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  
29098  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  
29099  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  
29100  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  

[29101 rows x 14 columns]

    pickup_dt – период с точностью до часа
    pickup_month – месяц
    borough – район Нью-Йорка, из которого был сделан заказ (5 районов + аэропорт)
    pickups – число поездок за период (час)
    hday – является ли день праздничным/выходным; Y - да,  N - нет
    spd – скорость ветра в милях в час
    vsb – видимость
    temp – температура в градусах Фаренгейта
    dewp – точка росы по Фаренгейту
    slp – давление
    pcp_01 – количество осадков за час
    pcp_06 – количество осадков за 6 часов
    pcp_24 – количество осадков за 24 часа
    sd – глубина снега в дюймах

In [4]:
taxi.shape

(29101, 14)

In [5]:
taxi.dtypes

pickup_dt        object
pickup_month     object
borough          object
pickups           int64
hday             object
spd             float64
vsb             float64
temp            float64
dewp            float64
slp             float64
pcp 01          float64
pcp 06          float64
pcp 24          float64
sd              float64
dtype: object

In [6]:
taxi = taxi.rename(columns={'pcp 01' : 'pcp_01', 'pcp 06' : 'pcp_06', 'pcp 24' : 'pcp_24'})

In [7]:
taxi.query("borough == 'Manhattan'").shape[0]

4343

In [8]:
taxi.value_counts('borough')

borough
Bronx            4343
Brooklyn         4343
EWR              4343
Manhattan        4343
Queens           4343
Staten Island    4343
dtype: int64

In [9]:
taxi.value_counts('borough').sum()

26058

In [10]:
pd.isnull(taxi).query('borough == True').shape[0]

3043

In [11]:
taxi.value_counts('borough').sum() + pd.isnull(taxi).query('borough == True').shape[0]

29101

Логика данных не нарушена, количество записей сохранилось

In [12]:
taxi

pickup_dt pickup_month        borough  pickups hday  spd  \
0      2015-01-01 01:00:00          Jan          Bronx      152    Y  5.0   
1      2015-01-01 01:00:00          Jan       Brooklyn     1519    Y  5.0   
2      2015-01-01 01:00:00          Jan            EWR        0    Y  5.0   
3      2015-01-01 01:00:00          Jan      Manhattan     5258    Y  5.0   
4      2015-01-01 01:00:00          Jan         Queens      405    Y  5.0   
...                    ...          ...            ...      ...  ...  ...   
29096  2015-06-30 23:00:00          Jun            EWR        0    N  7.0   
29097  2015-06-30 23:00:00          Jun      Manhattan     3828    N  7.0   
29098  2015-06-30 23:00:00          Jun         Queens      580    N  7.0   
29099  2015-06-30 23:00:00          Jun  Staten Island        0    N  7.0   
29100  2015-06-30 23:00:00          Jun            NaN        3    N  7.0   

        vsb  temp  dewp     slp  pcp_01  pcp_06  pcp_24   sd  
0      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
1      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
2      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
3      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
4      10.0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
...     ...   ...   ...     ...     ...     ...     ...  ...  
29096  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  
29097  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  
29098  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  
29099  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  
29100  10.0  75.0  65.0  1011.8     0.0     0.0     0.0  0.0  

[29101 rows x 14 columns]

#### Выяснить, из какого района было совершено наибольшее количество поездок за весь период

Найдем для начала суммарное количество всех поездок

In [13]:
taxi.aggregate({'pickups' : 'sum'})

pickups    14265773
dtype: int64

In [14]:
all_pickups = taxi.aggregate({'pickups' : 'sum'})[0]

In [15]:
all_pickups

14265773

In [16]:
taxi[['borough', 'pickups']].groupby('borough').aggregate({'pickups' : 'sum'}).sort_values('pickups', ascending=False)

pickups
borough                
Manhattan      10367841
Brooklyn        2321035
Queens          1343528
Bronx            220047
Staten Island      6957
EWR                 105

In [17]:
type(taxi[['borough', 'pickups']])

pandas.core.frame.DataFrame

In [18]:
taxi \
    .groupby('borough', as_index=True) \
    .aggregate({'pickups' : 'sum'}) \
    .sort_values('pickups', ascending=False) \
    .idxmax()

pickups    Manhattan
dtype: object

**Ответ** - Manhattan

In [19]:
taxi.groupby('borough', as_index=True).aggregate({'pickups' : 'sum'}).sort_values('pickups', ascending=False).sum()

pickups    14259513
dtype: int64

In [20]:
taxi \
.groupby('borough', as_index=True) \
.aggregate({'pickups' : 'sum'}) \
.sort_values('pickups', ascending=False) \
.sum()[0] - all_pickups

-6260

Суммарное значение по районам сопоставимо с количеством всех поездок, думаю, есть некоторая разница лишь из-за того, что в некоторых случаях район забора пассажира в таблице был пуст

In [21]:
taxi.query('borough.isnull()').head()

pickup_dt pickup_month borough  pickups hday  spd   vsb  temp  \
6   2015-01-01 01:00:00          Jan     NaN        4    Y  5.0  10.0  30.0   
13  2015-01-01 02:00:00          Jan     NaN       11    Y  3.0  10.0  30.0   
20  2015-01-01 03:00:00          Jan     NaN        1    Y  5.0  10.0  30.0   
27  2015-01-01 04:00:00          Jan     NaN        2    Y  5.0  10.0  29.0   
34  2015-01-01 05:00:00          Jan     NaN        3    Y  5.0  10.0  28.0   

    dewp     slp  pcp_01  pcp_06  pcp_24   sd  
6    7.0  1023.5     0.0     0.0     0.0  0.0  
13   6.0  1023.0     0.0     0.0     0.0  0.0  
20   8.0  1022.3     0.0     0.0     0.0  0.0  
27   9.0  1022.0     0.0     0.0     0.0  0.0  
34   9.0  1021.8     0.0     0.0     0.0  0.0

#### Сравните среднее число поездок, и выберите районы, из которых по праздникам в среднем поступает больше заказов, чем в обычные дни

In [22]:
taxi.describe()

pickups           spd           vsb          temp          dewp  \
count  29101.000000  29101.000000  29101.000000  29101.000000  29101.000000   
mean     490.215903      5.984924      8.818125     47.669042     30.823065   
std      995.649536      3.699007      2.442897     19.814969     21.283444   
min        0.000000      0.000000      0.000000      2.000000    -16.000000   
25%        1.000000      3.000000      9.100000     32.000000     14.000000   
50%       54.000000      6.000000     10.000000     46.000000     30.000000   
75%      449.000000      8.000000     10.000000     64.500000     50.000000   
max     7883.000000     21.000000     10.000000     89.000000     73.000000   

                slp        pcp_01        pcp_06        pcp_24            sd  
count  29101.000000  29101.000000  29101.000000  29101.000000  29101.000000  
mean    1017.817938      0.003830      0.026129      0.090464      2.529169  
std        7.768796      0.018933      0.093125      0.219402      4.520325  
min      991.400000      0.000000      0.000000      0.000000      0.000000  
25%     1012.500000      0.000000      0.000000      0.000000      0.000000  
50%     1018.200000      0.000000      0.000000      0.000000      0.000000  
75%     1022.900000      0.000000      0.000000      0.050000      2.958333  
max     1043.400000      0.280000      1.240000      2.100000     19.000000

In [23]:
taxi \
    .query('hday == "N"') \
    .groupby(['borough', 'hday']) \
    .aggregate({'pickups' : 'mean'})

pickups
borough       hday             
Bronx         N       50.771073
Brooklyn      N      534.727969
EWR           N        0.023467
Manhattan     N     2401.302921
Queens        N      308.899904
Staten Island N        1.606082

In [24]:
taxi \
    .query('hday == "Y"') \
    .groupby(['borough', 'hday']) \
    .aggregate({'pickups' : 'mean'})

pickups
borough       hday             
Bronx         Y       48.065868
Brooklyn      Y      527.011976
EWR           Y        0.041916
Manhattan     Y     2035.928144
Queens        Y      320.730539
Staten Island Y        1.497006

**Ответ** EWR, Queens

#### Для каждого района посчитайте число поездок по месяцам. Отсортируйте полученные значения по убыванию и сохраните результирующий датафрейм в pickups_by_mon_bor

Обратите внимание, что итоговый датасет должен состоять из 3-х колонок - pickup_month, borough, pickups.

Данные сохранены в переменную taxi.

In [25]:
taxi.head()

pickup_dt pickup_month    borough  pickups hday  spd   vsb  temp  \
0  2015-01-01 01:00:00          Jan      Bronx      152    Y  5.0  10.0  30.0   
1  2015-01-01 01:00:00          Jan   Brooklyn     1519    Y  5.0  10.0  30.0   
2  2015-01-01 01:00:00          Jan        EWR        0    Y  5.0  10.0  30.0   
3  2015-01-01 01:00:00          Jan  Manhattan     5258    Y  5.0  10.0  30.0   
4  2015-01-01 01:00:00          Jan     Queens      405    Y  5.0  10.0  30.0   

   dewp     slp  pcp_01  pcp_06  pcp_24   sd  
0   7.0  1023.5     0.0     0.0     0.0  0.0  
1   7.0  1023.5     0.0     0.0     0.0  0.0  
2   7.0  1023.5     0.0     0.0     0.0  0.0  
3   7.0  1023.5     0.0     0.0     0.0  0.0  
4   7.0  1023.5     0.0     0.0     0.0  0.0

Посмотрим, есть ли в таблице Null-месяца

In [26]:
taxi.query('pickup_month.isnull()').pickup_month.shape

(0,)

Может тогда Null-районы?

In [27]:
taxi.query('borough.isnull()').borough.shape

(3043,)

Отлично, замечено 3043 записи без районов, учту это в дальнейшем

In [28]:
taxi.groupby(['borough', 'pickup_month']).aggregate({'pickups' : 'sum'})

pickups
borough       pickup_month         
Bronx         Apr             34617
              Feb             28694
              Jan             22461
              Jun             49006
              Mar             32232
              May             53037
Brooklyn      Apr            378095
              Feb            328650
              Jan            309011
              Jun            482466
              Mar            346726
              May            476087
EWR           Apr                10
              Feb                14
              Jan                11
              Jun                29
              Mar                14
              May                27
Manhattan     Apr           1648278
              Feb           1718571
              Jan           1455543
              Jun           1995388
              Mar           1661261
              May           1888800
Queens        Apr            216857
              Feb            185695
              Jan            159211
              Jun            286311
              Mar            219561
              May            275893
Staten Island Apr              1068
              Feb               903
              Jan               821
              Jun              1673
              Mar               975
              May              1517

Сверим с общим количеством поездок

In [29]:
taxi \
    .groupby(['borough', 'pickup_month']) \
    .aggregate({'pickups' : 'sum'}) \
    .sum() - all_pickups

pickups   -6260
dtype: int64

Похоже на правду

In [30]:
taxi.query('borough.isnull()').pickups.sum()

6260

Вот оно и получается ;)

In [31]:
pickups_by_mon_bor = taxi \
                        .groupby(['pickup_month', 'borough'], as_index=False) \
                        .aggregate({'pickups' : 'sum'}) \
                        .sort_values('pickups', ascending=False)

In [32]:
pickups_by_mon_bor

pickup_month        borough  pickups
21          Jun      Manhattan  1995388
33          May      Manhattan  1888800
9           Feb      Manhattan  1718571
27          Mar      Manhattan  1661261
3           Apr      Manhattan  1648278
15          Jan      Manhattan  1455543
19          Jun       Brooklyn   482466
31          May       Brooklyn   476087
1           Apr       Brooklyn   378095
25          Mar       Brooklyn   346726
7           Feb       Brooklyn   328650
13          Jan       Brooklyn   309011
22          Jun         Queens   286311
34          May         Queens   275893
28          Mar         Queens   219561
4           Apr         Queens   216857
10          Feb         Queens   185695
16          Jan         Queens   159211
30          May          Bronx    53037
18          Jun          Bronx    49006
0           Apr          Bronx    34617
24          Mar          Bronx    32232
6           Feb          Bronx    28694
12          Jan          Bronx    22461
23          Jun  Staten Island     1673
35          May  Staten Island     1517
5           Apr  Staten Island     1068
29          Mar  Staten Island      975
11          Feb  Staten Island      903
17          Jan  Staten Island      821
20          Jun            EWR       29
32          May            EWR       27
26          Mar            EWR       14
8           Feb            EWR       14
14          Jan            EWR       11
2           Apr            EWR       10

#### Напишите функцию temp_to_celcius, которая получает на вход колонку с температурой в °F, и возвращает значения, переведенные в градусы Цельсия

Формула:

 $Celsius = \frac{(Fahrenheit-32) \cdot 5.0}{9.0}$

Надо написать только саму функцию. Так как урок посвящён pandas, то и всё содержимое функции должно быть выполнено средствами этой библиотеки - убедитесь, что на выходе она даёт pandas Series.

По идее это просто сделать с помощью векторизации? 

Что ж, попробую

In [33]:
def temp_to_celcius(col_name):
    """
        Функция перевода температуры из Фаренгейта в Цельсий
    
    Принимаемые параметры: pandas Series-колонку с температурой по Фаренгейту
    Выходные параметры: pandas Series
    """
    return (col_name - 32) * (5.0 / 9.0)

In [34]:
type(temp_to_celcius(taxi.temp))

pandas.core.series.Series